ddl/ddl_api.go

```diff
@@ -1809,12 +1807,16 @@ func BuildTableInfo(
 			return nil, dbterror.ErrUnsupportedClusteredSecondaryKey
 		}
 		if constr.Tp == ast.ConstraintForeignKey {
-			for _, fk := range tbInfo.ForeignKeys {
-				if fk.Name.L == strings.ToLower(constr.Name) {
-					return nil, infoschema.ErrCannotAddForeign
-				}
+			var fkName model.CIStr
+			if constr.Name != "" {
+				fkName = model.NewCIStr(constr.Name)
+			} else {
+				fkName = model.NewCIStr(fmt.Sprintf("fk_%v", tbInfo.MaxForeignKeyID+1))
 			}
-			fk, err := buildFKInfo(model.NewCIStr(constr.Name), constr.Keys, constr.Refer, cols, tbInfo)
+			if model.FindFKInfoByName(tbInfo.ForeignKeys, fkName.L) != nil {
+				return nil, infoschema.ErrCannotAddForeign
+			}
```

Comment:
The duplicate check has been implemented by `checkDuplicateConstraint`. Is this necessary?

Reply:
yes, since `checkConstraintNames` only check the `Constraint.Name`. If there are 2 constraints:
1. `Constraint` with specified name `fk_1`
2. `Constraint` without specified name, which will use generated name `fk_1`

then the name is duplicated, but  `checkDuplicateConstraint` can't check out this.